crates/holochain_zome_types/src/query.rs

```diff
@@ -118,116 +168,217 @@ pub struct ChainFork {
         self
     }
 
-    /// Filter on entry type
+    /// Filter on entry type.
     pub fn entry_type(mut self, entry_type: EntryType) -> Self {
         self.entry_type = Some(entry_type);
         self
     }
 
-    /// Filter on header type
+    /// Filter on entry hashes.
+    pub fn entry_hashes(mut self, entry_hashes: HashSet<EntryHash>) -> Self {
+        self.entry_hashes = Some(entry_hashes);
+        self
+    }
+
+    /// Filter on header type.
     pub fn header_type(mut self, header_type: HeaderType) -> Self {
         self.header_type = Some(header_type);
         self
     }
 
-    /// Include the entries in the ElementsVec that is returned
+    /// Include the entries in the ElementsVec that is returned.
     pub fn include_entries(mut self, include_entries: bool) -> Self {
         self.include_entries = include_entries;
         self
     }
 
-    /// Perform the boolean check which this filter represents
-    pub fn check(&self, header: &Header) -> bool {
-        let check_range = self
-            .sequence_range
-            .as_ref()
-            .map(|range| range.contains(&header.header_seq()))
-            .unwrap_or(true);
-        let check_header_type = self
-            .header_type
-            .as_ref()
-            .map(|header_type| header.header_type() == *header_type)
-            .unwrap_or(true);
-        let check_entry_type = self
-            .entry_type
-            .as_ref()
-            .map(|entry_type| {
-                header
-                    .entry_type()
-                    .map(|header_entry_type| *header_entry_type == *entry_type)
-                    .unwrap_or(false)
+    /// If the sequence range supports fork disambiguation, apply it to remove
+    /// headers that are not in the correct branch.
+    pub fn headers_without_forks(&self, headers: Vec<HeaderHashed>) -> Vec<HeaderHashed> {
+        match &self.sequence_range {
+            ChainQueryFilterSequenceRange::Unbounded => headers,
+            ChainQueryFilterSequenceRange::HeaderSeqRange(start, end) => headers
+                .into_iter()
+                .filter(|header| *start <= header.header_seq() && header.header_seq() <= *end)
+                .collect(),
+            ChainQueryFilterSequenceRange::HeaderHashRange(start, end) => {
+                let mut header_hashmap = headers
+                    .iter()
+                    .map(|header| (header.as_hash().clone(), header))
+                    .collect::<HashMap<HeaderHash, &HeaderHashed>>();
+                let mut filtered_headers = Vec::new();
+                let mut maybe_next_header = header_hashmap.remove(end);
+                while let Some(next_header) = maybe_next_header {
+                    maybe_next_header = next_header
+                        .as_content()
+                        .prev_header()
+                        .and_then(|prev_header| header_hashmap.remove(prev_header));
+                    filtered_headers.push(next_header.clone());
```

Comment:
If you make the above hashmap own the hashed header you don't need to clone the whole header here every time